internal/crossdock/client/oneway/protobuf.go

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package oneway
+
+import (
+	"context"
+
+	"go.uber.org/yarpc"
+	"go.uber.org/yarpc/internal/crossdock/crossdockpb"
+
+	"github.com/crossdock/crossdock-go"
+)
+
+// Protobuf starts an http oneway run using Protobuf encoding
+func Protobuf(t crossdock.T, dispatcher *yarpc.Dispatcher, serverCalledBack <-chan []byte, callBackAddr string) {
+	assert := crossdock.Assert(t)
+	fatals := crossdock.Fatals(t)
+
+	client := crossdockpb.NewOnewayYarpcClient(dispatcher.ClientConfig("oneway-server"))
+	token := getRandomID()
+
+	ack, err := client.Echo(context.Background(), &crossdockpb.Token{Value: token}, yarpc.WithHeader("callBackAddr", callBackAddr))
+
+	// ensure channel hasn't been filled yet
+	select {
+	case <-serverCalledBack:
+		fatals.FailNow("oneway protobuf test failed", "client waited for server to fill channel")
+	default:
+	}
+
+	fatals.NoError(err, "call to Oneway::echo failed: %v", err)
+	fatals.NotNil(ack, "ack is nil")
+
+	serverToken := <-serverCalledBack
```

Comment:
I realize that this is virtually identical to the existing Thrift code, but I'm not grokking something.

On line 44, we fail the test if we can read from `serverCalledBack`. Between line 44 and here, there's no explicit synchronization - in fact, there's nothing but a few asserts, which should take microseconds at most. On this line, though, we *want* to rea from the channel. Why are we guaranteed that the first read fails, while this one succeeds? Should we move lines 43-47 before the `client.Echo` call?